### PR TITLE
add document360 connector

### PR DESCRIPTION
**Pull request by nlp8899**

### CODE REVIEW

File: backend/danswer/connectors/document360/connector.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
+import requests
+from datetime import datetime, timezone
+from typing import Any, Generator, Optional, List, Union
```
Comment (yuhongsun96) [inline backend/danswer/connectors/document360/connector.py]: Looks like you might be a consistent contributor! Happy to see that!

Please follow the [CONTRIBUTING](https://github.com/danswer-ai/danswer/blob/main/CONTRIBUTING.md) guide to make sure your code is linted and type checking passes

Linter should take care of restructuring the imports. The Mypy issues need to be addressed manually though. Please address those, thanks!

File: backend/danswer/connectors/document360/connector.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+        }
+
+        response = requests.get(f"{DOCUMENT360_API_BASE_URL}/{endpoint}", headers=headers, params=params)
+        response.raise_for_status()
+
+        return response.json()['data']
+
+    def _get_workspace_id_by_name(self):
+        projects = self._make_request("ProjectVersions")
+        return next((project['id'] for project in projects if project["version_code_name"] == self.workspace), None)
```
Comment (yuhongsun96) [inline backend/danswer/connectors/document360/connector.py]: Is the user provides an invalid workspace, the connector should fail to alert the user that something has gone wrong. The error message for the connector is passed to the frontend so we should add a informative message

File: backend/danswer/connectors/document360/connector.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+    
+    def _get_articles_with_category(self, workspace_id):
+        all_categories = self._make_request(f"ProjectVersions/{workspace_id}/categories")
+        articles_with_category = []
+
+        for category in all_categories:
+            if category["name"] in self.categories or "ALL" in self.categories:
```
Comment (yuhongsun96) [inline backend/danswer/connectors/document360/connector.py]: Let's make categories list[str] or None. If None then it should pull from all categories. It's a bit of an edge case but we don't want any potential weird match with "ALL" as a category name

File: web/src/app/admin/connectors/document360/page.tsx [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+              <TrashIcon />
+            </button>
+          </div>
+        </>
+      ) : (
+        <>
+          <p className="text-sm mb-4">
+            To use the Document360 connector, you must first provide the API token 
+            and portal ID corresponding to your Document360 setup. For more details, 
+            see the <a className="text-blue-500" href="https://docs.document360.io/">official Document360 documentation</a>.
```
Comment (yuhongsun96) [inline web/src/app/admin/connectors/document360/page.tsx]: https://apidocs.document360.com/apidocs/api-token

Can we link to this instead?

File: web/src/app/admin/connectors/document360/page.tsx [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+            >
+              <TrashIcon />
+            </button>
+          </div>
+        </>
+      ) : (
+        <>
+          <p className="text-sm mb-4">
+            To use the Document360 connector, you must first provide the API token 
+            and portal ID corresponding to your Document360 setup. For more details,
```
Comment (yuhongsun96) [inline web/src/app/admin/connectors/document360/page.tsx:96]: It's not immediately obvious what the portal ID is, should we link to the Danswer docs page for this?

File: backend/danswer/connectors/document360/connector.py [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+                        "id": article["id"],
+                        "category_name": category["name"]
+                    })
+                for child_category in category["child_categories"]:
+                    for article in child_category["articles"]:
+                        articles_with_category.append({
+                            "id": article["id"],
+                            "category_name": child_category["name"]
+                        })
+        return articles_with_category
```
Comment (yuhongsun96) [inline backend/danswer/connectors/document360/connector.py:91]: What happens if the user has a large number of articles? Should the request be paginated/batched?

File: backend/danswer/connectors/document360/connector.py [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+            article_details = self._make_request(f"Articles/{article['id']}", {"langCode": "en"})
+            
+            updated_at = datetime.strptime(article_details["modified_at"], "%Y-%m-%dT%H:%M:%S.%fZ").replace(tzinfo=None)
+            if start is not None and updated_at < start:
+                continue
+            if end is not None and updated_at > end:
+                continue
+            
+            doc_link = f"{DOCUMENT360_BASE_URL}/{self.portal_id}/document/v1/article/{article['id']}"
+            doc_text = f"workspace: {self.workspace}\n category: {article['category_name']}\n article: {article_details['title']} - {article_details.get('description', '')}"
```
Comment (yuhongsun96) [inline backend/danswer/connectors/document360/connector.py]: We definitely want to pull content and html_content and if either of those exist, parse those into the actual contents of the article

**Review by yuhongsun96 — CHANGES_REQUESTED** [review]
Great work! We do need a few changes though:
- Biggest one is the contents of the article needs to be brought in and parsed
- A few smaller edge case/minor fixes

Also important, could you create a PR for docs as well? Otherwise people won't know how to use your connector: https://github.com/danswer-ai/danswer-docs

It's a no-code type doc, just markdown, I can help you set it up if you'd like.

File: backend/danswer/connectors/document360/connector.py (hunk 1 above)
Comment (nlp8899) [reply backend/danswer/connectors/document360/connector.py]: Okay, I fixed type checking issue.

File: backend/danswer/connectors/document360/connector.py (hunk 2 above)
Comment (nlp8899) [reply backend/danswer/connectors/document360/connector.py]: Okay, I will handle this error.

File: backend/danswer/connectors/document360/connector.py (hunk 3 above)
Comment (nlp8899) [reply backend/danswer/connectors/document360/connector.py]: I agree with you. I update for this one.

File: web/src/app/admin/connectors/document360/page.tsx (hunk 4 above)
Comment (nlp8899) [reply web/src/app/admin/connectors/document360/page.tsx]: Okay, I will. great catch.

File: web/src/app/admin/connectors/document360/page.tsx (hunk 5 above)
Comment (nlp8899) [reply web/src/app/admin/connectors/document360/page.tsx:96]: Portal ID is only used to get article link. Each document360 portal dashboard has its own portal id. (it is unique id for each api token)

File: backend/danswer/connectors/document360/connector.py (hunk 6 above)
Comment (nlp8899) [reply backend/danswer/connectors/document360/connector.py:91]: Yes, I feel it should be batched for large number of articles. I will handle it.

**Pull request closed without merging**